pkgs/operator/state.go

```diff
@@ -531,3 +509,155 @@ func (s *Switch) GetSecretShare(id [24]byte) (*kyber_dkg.DistKeyShare, error) {
 	}
 	return &kyber_dkg.DistKeyShare{Share: &share.PriShare{V: secretPoint, I: privShare.Share.I}, Commits: coefs}, nil
 }
+
+func (s *Switch) Pong(incMsg *wire.SignedTransport) ([]byte, error) {
+	operatorID, err := s.VerifyIncomingMessage(incMsg)
+	if err != nil {
+		return nil, err
+	}
+	pong := &wire.Pong{
+		ID:     operatorID,
+		PubKey: s.PubKeyBytes,
+	}
+	return s.MarshallAndSign(pong, wire.PongMessageType, operatorID, [24]byte{})
+}
+
+func (s *Switch) SaveResultData(incMsg *wire.SignedTransport) error {
+	resData := &wire.ResultData{}
+	if err := resData.UnmarshalSSZ(incMsg.Message.Data); err != nil {
+		return err
+	}
+	_, err := s.VerifyIncomingMessage(incMsg)
+	if err != nil {
+		return err
+	}
+	// store deposit result
+	timestamp := time.Now().Format(time.RFC3339Nano)
+	dir := fmt.Sprintf("%s/ceremony-%s", cli_utils.OutputPath, timestamp)
+	if _, err := os.Stat(dir); os.IsNotExist(err) {
+		err = os.Mkdir(dir, os.ModePerm)
+		if err != nil {
+			return err
+		}
+	}
+	var depJson *initiator.DepositDataJson
+	err = json.Unmarshal(resData.DepositData, &depJson)
+	if err != nil {
+		return err
+	}
+	err = cli_utils.WriteDepositResult(depJson, dir)
+	if err != nil {
+		return err
+	}
+	// store keyshares result
+	var ksJson *initiator.KeyShares
+	err = json.Unmarshal(resData.KeysharesData, &ksJson)
+	if err != nil {
+		return err
+	}
+	err = cli_utils.WriteKeysharesResult(ksJson, dir, incMsg.Message.Identifier)
+	if err != nil {
+		return err
+	}
+	// store instance ID
+	err = cli_utils.WriteInstanceID(dir, incMsg.Message.Identifier)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (s *Switch) VerifyIncomingMessage(incMsg *wire.SignedTransport) (uint64, error) {
+	var initiatorPubKey *rsa.PublicKey
+	var ops []*wire.Operator
+	var err error
+	switch incMsg.Message.Type {
+	case wire.PingMessageType:
+		ping := &wire.Ping{}
+		if err := ping.UnmarshalSSZ(incMsg.Message.Data); err != nil {
+			return 0, err
+		}
+		// Check that incoming init message signature is valid
+		initiatorPubKey, err = crypto.ParseRSAPubkey(ping.InitiatorPublicKey)
+		if err != nil {
+			return 0, err
+		}
+		ops = ping.Operators
+	case wire.ResultMessageType:
+		resData := &wire.ResultData{}
+		if err := resData.UnmarshalSSZ(incMsg.Message.Data); err != nil {
+			return 0, err
+		}
+		// Check that incoming init message signature is valid
+		initiatorPubKey, err = crypto.ParseRSAPubkey(resData.InitiatorPublicKey)
+		if err != nil {
+			return 0, err
+		}
+		ops = resData.Operators
+	}
+	err = s.VerifySig(incMsg, initiatorPubKey)
+	if err != nil {
+		return 0, err
+	}
+	operatorID, err := GetOperatorID(ops, s.PubKeyBytes)
```

Comment:
I don't see a point in verifying signature of message only against a public key included in the message itself. are you sure this is what you meant to do here?

Reply:
We follow this logic everywhere, I just didnt want to deviate

Reply:
If we do this everywhere then this is really bad. on the operators side we should validate the signature using the initiator pubkey we received on init and not in every message, this doesn't add any security like this.
also on the initiator (or on operators when they extract every operator message from the combined message) side we should validate the messages using the pubkey of an operator that claims to have signed the message. we should have them from beforehand in the INIT . in fact its not a must to have pubkey inside the message, an ID will be enough and then we verify we the pubkey we know belongs to that ID.

Reply:
- we can move initiator pub to instance at operator side, then we will have a relation between instance ID and initiator`s pub key
- at initiator side we have operator pubs so able to verify operator message sigs

Yes, we can get rid of pub keys at messages.

Reply:
Update: 
in case of Result messages we have already verification of ID + pub. Ill reuse it
In case of ping/pong for health check ill keep it, as we dont have instance related to it.

Reply:
I still think validating the key that is inside the message without comparing it to anything doesn't add any security, anyone can create a valid key and sign with it.

Reply:
Okay, then do we need any security doing health_check? Probably no for initiator side. I changed to GET requests and verification of operator sigs